File: maze/html/src/maze.js
class Coord {
    constructor(x, y) {
        if (!Number.isInteger(x) || !Number.isInteger(y)) {
            throw new Error("Coordinates must be integers");
        }
        if (x < 0 || y < 0) {
            throw new Error("Coordinates must be non-negative");
        }
        this.x = x;
        this.y = y;
    }

    getCoord() {
        return [this.x, this.y];
    }

    getX() {
        return this.x;
    }

    getY() {
        return this.y;
    }

    getTop() {
        if (this.y === 0) {
            return false;
        }
        return new Coord(this.x, this.y - 1);
    }

    getRight() {
        return new Coord(this.x + 1, this.y);
    }

    getBottom() {
        return new Coord(this.x, this.y + 1);
    }

    getLeft() {
        if (this.x === 0) {
            return false;
        }
        return new Coord(this.x - 1, this.y);
    }

    getDirection(coord) {
        const top = this.getTop();
        if (top && coord.equals(top)) {
            return "t";
        }
        if (coord.equals(this.getRight())) {
            return "r";
        }
        if (coord.equals(this.getBottom())) {
            return "b";
        }
        const left = this.getLeft();
        if (left && coord.equals(left)) {
            return "l";
        }
        return "n";
    }

    toString() {
        return `${this.x},${this.y}`;
    }

    equals(other) {
        return this.x === other.getX() && this.y === other.getY();
    }

    hashCode() {
        return `${this.x},${this.y}`;
    }
}

class Cell {
    constructor() {
        this.rightWall = true;
        this.bottomWall = true;
        this.explored = false;
        this.pathed = false;
    }

    getRightWall() {
        return this.rightWall;
    }

    getBottomWall() {
        return this.bottomWall;
    }

    breakRightWall() {
        this.rightWall = false;
    }

    breakBottomWall() {
        this.bottomWall = false;
    }

    getExplored() {
        return this.explored;
    }

    setExplored() {
        this.explored = true;
    }

    unsetExplored() {
        this.explored = false;
    }

    getPathed() {
        return this.pathed;
    }

    setPathed() {
        this.pathed = true;
    }

    toString() {
        let res = "";
        res += this.bottomWall ? "_" : " ";
        res += this.rightWall ? "|" : " ";
        return res;
    }
}

class Maze {
    constructor(length, height) {
        if (!Number.isInteger(length) || !Number.isInteger(height)) {
            throw new Error("Length and height must be integers");
        }
        if (length <= 0 || height <= 0) {
            throw new Error("Length and height must be positive");
        }
        this.length = length;
        this.height = height;
        this.cells = new Map();
        this.outs = [];

        for (let y = 0; y < height; y++) {
            for (let x = 0; x < length; x++) {
                const coord = new Coord(x, y);
                this.cells.set(coord.hashCode(), new Cell());
            }
        }
    }

    getCell(coord) {
        return this.cells.get(coord.hashCode());
    }

    checkSurround(coord) {
        const res = [];
        const top = coord.getTop();
        if (top && !this.getCell(top).getPathed()) {
            res.push(top);
        }
        if (coord.getX() < this.length - 1 && !this.getCell(coord.getRight()).getPathed()) {
            res.push(coord.getRight());
        }
        if (coord.getY() < this.height - 1 && !this.getCell(coord.getBottom()).getPathed()) {
            res.push(coord.getBottom());
        }
        const left = coord.getLeft();
        if (left && !this.getCell(left).getPathed()) {
            res.push(left);
        }
        return res;
    }

    backtrack(coord) {
        const top = coord.getTop();
        if (top && this.getCell(top).getExplored() && !this.getCell(top).getBottomWall()) {
            return top;
        }
        if (coord.getX() < this.length - 1 && this.getCell(coord.getRight()).getExplored() && !this.getCell(coord).getRightWall()) {
            return coord.getRight();
        }
        if (coord.getY() < this.height - 1 && this.getCell(coord.getBottom()).getExplored() && !this.getCell(coord).getBottomWall()) {
            return coord.getBottom();
        }
        const left = coord.getLeft();
        if (left && this.getCell(left).getExplored() && !this.getCell(left).getRightWall()) {
            return left;
        }
        return null;
    }

    generate() {
        const edges = this.getEdgeCoordinates();
        this.outs = this.getRandomSample(edges, 2);
        
        let coord = this.outs[0];
        let explored = 1;
        this.getCell(coord).setExplored();
        this.getCell(coord).setPathed();

        while (explored < this.length * this.height) {
            let possibilities;
            if (coord.equals(this.outs[1])) {
                possibilities = [];
            } else {
                possibilities = this.checkSurround(coord);
            }

            if (possibilities.length === 0) {
                this.getCell(coord).unsetExplored();
                coord = this.backtrack(coord);
                if (coord === null) {
                    break;
                }
            } else {
                const newCoord = possibilities[Math.floor(Math.random() * possibilities.length)];
                const direction = coord.getDirection(newCoord);

                if (direction === "t") {
                    this.getCell(newCoord).breakBottomWall();
                } else if (direction === "r") {
                    this.getCell(coord).breakRightWall();
                } else if (direction === "b") {
                    this.getCell(coord).breakBottomWall();
                } else if (direction === "l") {
                    this.getCell(newCoord).breakRightWall();
                }

                coord = newCoord;
                this.getCell(coord).setExplored();
                this.getCell(coord).setPathed();
                explored++;
            }
        }
    }

    getEdgeCoordinates() {
        const edges = [];
        for (let x = 0; x < this.length; x++) {
            edges.push(new Coord(x, 0));
            edges.push(new Coord(x, this.height - 1));
        }
        for (let y = 1; y < this.height - 1; y++) {
            edges.push(new Coord(0, y));
            edges.push(new Coord(this.length - 1, y));
        }
        return edges;
    }

    getRandomSample(arr, k) {
        const shuffled = [...arr].sort(() => Math.random() - 0.5);
        return shuffled.slice(0, k);
    }

    getCells() {
        return this.cells;
    }

    solve() {
        const start = this.outs[0];
        const end = this.outs[1];
        
        const queue = [[start, [start]]];
        const visited = new Set();
        visited.add(start.hashCode());

        while (queue.length > 0) {
            const [current, path] = queue.shift();

            if (current.equals(end)) {
                return path;
            }

            const neighbors = [
                { coord: current.getTop(), check: () => current.getTop() && !this.getCell(current.getTop()).getBottomWall() },
                { coord: current.getRight(), check: () => !this.getCell(current).getRightWall() },
                { coord: current.getBottom(), check: () => !this.getCell(current).getBottomWall() },
                { coord: current.getLeft(), check: () => current.getLeft() && !this.getCell(current.getLeft()).getRightWall() }
            ];

            for (const neighbor of neighbors) {
                if (neighbor.coord && neighbor.check() && !visited.has(neighbor.coord.hashCode())) {
                    visited.add(neighbor.coord.hashCode());
                    queue.push([neighbor.coord, [...path, neighbor.coord]]);
                }
            }
        }

        return [];
    }

    toString() {
        let res = " " + "_ ".repeat(this.length) + "\n";
        for (let y = 0; y < this.height; y++) {
            res += "|";
            for (let x = 0; x < this.length; x++) {
                const coord = new Coord(x, y);
                res += this.getCell(coord).toString();
            }
            res += "\n";
        }
        res += "in: " + this.outs[0].toString() + "\n";
        res += "out: " + this.outs[1].toString() + "\n";
        return res;
    }
}

File: maze/python/maze.py
import random

#top, right, bottom, left

class Coord:
    
    def __init__(self,x,y):
        assert type(x) == int
        assert type(y) == int
        assert x >= 0
        assert y >= 0
        self.x = x
        self.y = y
        
    def getCoord(self):
        return (self.x,self.y)
    
    def getX(self):
        return self.x
    
    def getY(self):
        return self.y
    
    def getTop(self):
        if(self.y == 0):
            return False
        return Coord(self.x,self.y-1)
    
    def getRight(self):
        return Coord(self.x+1,self.y)
    
    def getBottom(self):
        return Coord(self.x,self.y+1)
    
    def getLeft(self):
        if(self.x == 0):
            return False
        return Coord(self.x-1,self.y)
    
    def getDirection(self,coord):
        if(self.getTop() and coord == self.getTop()):
            return "t"
        if(coord == self.getRight()):
            return "r"
        if(coord == self.getBottom()):
            return "b"
        if(self.getLeft() and coord == self.getLeft()):
            return "l"
        return "n"
    
    def __str__(self):
        return(str(self.x)+","+str(self.y))

    def __eq__(self, other):
        return self.x == other.getX() and self.y == other.getY()
    
    def __hash__(self):
        return hash((self.x, self.y))


class Cell:
    
    def __init__(self):
        self.rightWall = True
        self.bottomWall = True
        self.explored = False
        self.pathed = False
    
    def getWalls(self):
        return self.walls
    
    def getRightWall(self):
        return self.rightWall
    
    def getBottomWall(self):
        return self.bottomWall
    
    def breakRightWall(self):
        self.rightWall = False
        
    def breakBottomWall(self):
        self.bottomWall = False

    def getExplored(self):
        return self.explored
    
    def setExplored(self):
        self.explored = True
        
    def unsetExplored(self):
        self.explored = False
        
    def getPathed(self):
        return self.pathed
        
    def setPathed(self):
        self.pathed = True
        
    def __str__(self):
        res = ""
        if(self.bottomWall):
            res += "_"
        else:
            res += " "
        if(self.rightWall):
            res += "|"
        else:
            res += " "
        return res
        
        
class Maze:
    
    def __init__(self,length,height):
        assert type(length) == int
        assert type(height) == int
        assert length > 0
        assert height > 0
        self.length = length
        self.height = height
        self.cells = dict()
        self.outs = []
        for y in range(height):
            for x in range(length):
                self.cells[Coord(x,y)] = Cell()
                
    def getCell(self,coord):
        assert type(coord) == Coord
        return self.cells[coord]
        
    def checkSurround(self,coord):
        res = []
        if(coord.getY()>0 and not self.cells[coord.getTop()].getPathed()):
            res.append(coord.getTop())
        if(coord.getX()<self.length-1 and not self.cells[coord.getRight()].getPathed()):
            res.append(coord.getRight())
        if(coord.getY()<self.height-1 and not self.cells[coord.getBottom()].getPathed()):
            res.append(coord.getBottom())
        if(coord.getX()>0 and not self.cells[coord.getLeft()].getPathed()):
            res.append(coord.getLeft())
        return res
        
    def backtrack(self,coord):
        if(coord.getY()>0 and self.cells[coord.getTop()].getExplored() and not self.cells[coord.getTop()].getBottomWall()):
            return coord.getTop()
        if(coord.getX()<self.length-1 and self.cells[coord.getRight()].getExplored() and not self.cells[coord].getRightWall()):
            return coord.getRight()
        if(coord.getY()<self.height-1 and self.cells[coord.getBottom()].getExplored() and not self.cells[coord].getBottomWall()):
            return coord.getBottom()
        if(coord.getX()>0 and self.cells[coord.getLeft()].getExplored() and not self.cells[coord.getLeft()].getRightWall()):
            return coord.getLeft()
        
    def generate(self):
        edges = self.getEdgeCoordinates()
        outs = random.sample(edges, 2)
        self.outs = outs
        coord = outs[0]
        explored = 1
        self.cells[coord].setExplored()
        self.cells[coord].setPathed()
        while explored < self.length*self.height:
            if(coord.getCoord() == outs[1].getCoord()):
                possibilities = []
            else:
                possibilities = self.checkSurround(coord)
            if(possibilities == []):
                self.cells[coord].unsetExplored()
                coord = self.backtrack(coord)
            else:
                newCoord = random.choice(possibilities)
                direction = coord.getDirection(newCoord)
                if(direction == "t"):
                    self.cells[newCoord].breakBottomWall()
                elif(direction == "r"):
                    self.cells[coord].breakRightWall()
                elif(direction == "b"):
                    self.cells[coord].breakBottomWall()
                elif(direction == "l"):
                    self.cells[newCoord].breakRightWall()
                coord = newCoord
                self.cells[coord].setExplored()
                self.cells[coord].setPathed()
                explored += 1
                
    def __str__(self):
        res = (" _"*(self.length))+"\n"
        for y in range(self.height):
            res+="|"
            for x in range(self.length):
                res+= str(self.cells[Coord(x,y)])
            res += "\n"
        res += "in: "+str(self.outs[0])+"\n"
        res += "out: "+str(self.outs[1])+"\n"
        return res

    def getCells(self):
        return self.cells
    
    def getEdgeCoordinates(self):
        edges = []
        for x in range(self.length):
            edges.append(Coord(x, 0))  # top edge
            edges.append(Coord(x, self.height-1))  # bottom edge
        for y in range(1, self.height-1):
            edges.append(Coord(0, y))  # left edge
            edges.append(Coord(self.length-1, y))  # right edge
        return edges
    
if __name__ == '__main__':
    lab = Maze(15,15)
    lab.generate()
    with open("demo.txt", "w") as f:
        f.write(str(lab))